python/cudf/cudf/core/frame.py

```diff
@@ -1236,6 +1236,19 @@ def replace(self, to_replace, replacement):
 
         for name, col in copy_data.items():
             if not (to_replace is None and replacement is None):
+                # For Index replace if replacement has string change
+                # thecolumn type to string so mixed replacements
+                # will select the proper column type and function.
+                if (
+                    not hasattr(self, "index")
+                    and any(type(item) == str for item in replacement)
+                    and type(self) != cudf.core.index.StringIndex
+                ):
+                    raise NotImplementedError(
+                        "Implicit conversion of index to "
+                        "mixed type is not yet supported."
+                    )
```

Comment:
`replacement` can be a cudf backed object here that we can't iterate over

additionally, we shouldn't need any Index specific logic in here, this is just a collection of columns that we're operation on at this point

Reply:
We should handle iterating over the values in the `rename` function instead where we know we're dealing with Python dictionaries / lists.